### PR TITLE
Modify MoleculeGraphData  and add CGCNN featurizer

### PR DESCRIPTION
**Pull request by nd-02110114**
## What I did

- [x] Refactor `MoleculeGraphData ` -> `GraphData`
  - Remove targets attributes (featurizer doesn't take target)
  - Rename class name
- [x] Refactor `StructureGraphFeaturizer ` -> `CGCNNFeaturizer` using `GraphData` class
- [x] Split `materials_featurizers.py`

### CODE REVIEW

File: deepchem/feat/material_featurizers/cgcnn_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import numpy as np
+from typing import Tuple
+
+from deepchem.utils import download_url, get_data_dir
+from deepchem.utils.typing import PymatgenStructure
+from deepchem.feat import MaterialStructureFeaturizer
+from deepchem.feat.graph_data import GraphData
+
+# FIXME: it is better to add this json to DeepChem AWS
+ATOM_INIT_JSON_URL = 'https://raw.githubusercontent.com/txie-93/cgcnn/master/data/sample-regression/atom_init.json'
```
Comment (nd-02110114) [inline deepchem/feat/material_featurizers/cgcnn_featurizer.py:12]: Could I upload this json file....?

File: deepchem/feat/graph_data.py [hunk 2]
```diff
@@ -2,73 +2,69 @@
...
-  node_features : np.ndarray
+  node_features: np.ndarray
     Node feature matrix with shape [num_nodes, num_node_features]
-  edge_index : np.ndarray, dtype int
+  edge_index: np.ndarray, dtype int
     Graph connectivity in COO format with shape [2, num_edges]
-  targets : np.ndarray
-    Graph or node targets with arbitrary shape
-  edge_features : np.ndarray, optional (default None)
+  edge_features: np.ndarray, optional (default None)
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:17]: One generalization we might want to consider is adding `pair_features`. The `WeaveFeaturizer` class has pair features for all pairs of atoms, even those which aren't connected by an edge

File: deepchem/feat/graph_data.py [hunk 3]
```diff
@@ -2,73 +2,69 @@
 import numpy as np
 
 
-class MoleculeGraphData:
-  """MoleculeGraphData class
+class GraphData:
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:5]: +1 for the new name!

File: deepchem/feat/graph_data.py [hunk 4]
```diff
@@ -2,73 +2,69 @@
...
+  """GraphData class
 
   This data class is almost same as `torch_geometric.data.Data 
   <https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data>`_.
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:10]: Could you add an example showing how to instantiate a `GraphData` class and extract its attributes?

File: deepchem/feat/graph_data.py [hunk 5]
```diff
@@ -77,75 +73,77 @@ def __init__(
...
     self.edge_features = edge_features
     self.graph_features = graph_features
-    self.targets = targets
     self.num_nodes, self.num_node_features = self.node_features.shape
     self.num_edges = edge_index.shape[1]
     if self.node_features is not None:
       self.num_edge_features = self.edge_features.shape[1]
 
-  def to_pyg_data(self):
+  def to_pyg_data(self, target: np.ndarray):
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:81]: Perhaps we should just go ahead and add a test for this. We'll have to add pytorch-geometric to `.travis.yml` but that might be worth the additional work to ensure correctness

File: deepchem/feat/graph_data.py [hunk 6]
```diff
@@ -77,75 +73,77 @@ def __init__(
...
+      y=torch.from_numpy(target),
     )
 
 
-class BatchMoleculeGraphData(MoleculeGraphData):
-  """Batch MoleculeGraphData class
-  
+class BatchGraphData(GraphData):
+  """Batch GraphData class
+
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:112]: Could you add an example for this class's usage?

File: deepchem/feat/material_featurizers/cgcnn_featurizer.py (hunk 1 above)
Comment (rbharath) [reply deepchem/feat/material_featurizers/cgcnn_featurizer.py:12]: Yes absolutely! I'll try to get AWS permissions fixed for you by tomorrow

File: deepchem/feat/material_featurizers/cgcnn_featurizer.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  representation including atom features and bond features (neighbor
+  distances). Neighbors are determined by searching in a sphere around
+  atoms in the unit cell. A Gaussian filter is applied to neighbor distances.
+  All units are in angstrom.
+
+  This featurizer requires the optional dependency pymatgen. It may
+  be useful when 3D coordinates are available and when using graph
+  network models and crystal graph convolutional networks.
+
+  See [1]_ for more details.
```
Comment (rbharath) [inline deepchem/feat/material_featurizers/cgcnn_featurizer.py:30]: Could you add an example showing how to invoke this featurizer?

File: deepchem/feat/material_featurizers/element_property_fingerprint.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  based on elemental stoichiometry. E.g., the average electronegativity
+  of atoms in a crystal structure. The chemical fingerprint is a
+  vector of these statistics. For a full list of properties and statistics,
+  see ``matminer.featurizers.composition.ElementProperty(data_source).feature_labels()``.
+
+  This featurizer requires the optional dependencies pymatgen and
+  matminer. It may be useful when only crystal compositions are available
+  (and not 3D coordinates).
+
+  See references [1]_ [2]_ [3]_ [4]_ for more details.
```
Comment (rbharath) [inline deepchem/feat/material_featurizers/element_property_fingerprint.py:22]: Could you add a usage example to this class?

File: deepchem/feat/material_featurizers/sine_coulomb_matrix.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  Features are flattened into a vector of matrix eigenvalues by default
+  for ML-readiness. To ensure that all feature vectors are equal
+  length, the maximum number of atoms (eigenvalues) in the input
+  dataset must be specified.
+
+  This featurizer requires the optional dependencies pymatgen and
+  matminer. It may be useful when crystal structures with 3D coordinates
+  are available.
+
+  See [1]_ for more details.
```
Comment (rbharath) [inline deepchem/feat/material_featurizers/sine_coulomb_matrix.py:28]: Could you add a usage example to this class?

File: deepchem/utils/typing.py [hunk 10]
```diff
@@ -16,6 +16,10 @@
 # The shape of a NumPy array
 Shape = Tuple[int, ...]
 
-# type of RDKit Mol object
+# type of RDKit object
 RDKitMol = Any
 RDKitAtom = Any
+
+# type of Pymatgen object
```
Comment (rbharath) [inline deepchem/utils/typing.py:23]: +1 for these type labels! This is a nice way to get around the optional imports hampering typing :)

File: deepchem/feat/graph_data.py (hunk 2 above)
Comment (nd-02110114) [reply deepchem/feat/graph_data.py:17]: I seem that we can define the pair features using the edge_idx and edge_features, but I'm not deeply understanding WeaveNet Implementation. 

I think it is better to add the attribute when we start to building the WeaveModel using new graph data.

Comment (rbharath) [reply deepchem/feat/graph_data.py:17]: Sounds good! Probably better to generalize with a concrete use case in hand

File: deepchem/feat/graph_data.py [hunk 11]
```diff
@@ -2,73 +2,69 @@
...
+  """GraphData class
 
-  This data class is almost same as `torch_geometric.data.Data 
+  This data class is almost same as `torch_geometric.data.Data
   <https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data>`_.
 
   Attributes
   ----------
-  node_features : np.ndarray
+  node_features: np.ndarray
```
Comment (rbharath) [inline deepchem/feat/graph_data.py:13]: As an additional suggestion, should we just go ahead and add a converter to a DGL graph? (https://docs.dgl.ai/en/0.4.x/api/python/graph.html). That would make our interop with PyG/DGL pretty seamless once we finish converting out models over to `GraphData`

File: scripts/install_deepchem_conda.sh [hunk 12]
```diff
@@ -22,4 +22,27 @@ else
...
+list=(`cat $PWD/requirements-pyg.txt | xargs`)
+for pkg in "${list[@]}" ; do
+    pkg=`echo ${pkg} | sed -e "s/[\r\n]\+//g"`
+    if [[ $pkg =~ torch-geometric ]];
+    then
+        pip install ${pkg}
+    else
+        pip install ${pkg}+cpu -f https://pytorch-geometric.com/whl/torch-${TORCH}.html
+    fi
+done
```
Comment (nd-02110114) [inline scripts/install_deepchem_conda.sh:48]: I add these lines for PyTorch and  PyTorch Geometric,  but I couldn't write for  PowerShell script...

**Comment by nd-02110114** [comment]
This PR is ready to a second review!
The CI failed in Windows, but the error is not related to this PR.
The error is related to `make_pytorch_dataset` method, so I will try to fix in an another PR.

**Comment by peastman** [comment]
Would it make sense to represent indices as `np.int32` instead of `np.int64`?  It might improve performance.

**Comment by rbharath** [comment]
I think our PyTorch Geometric tests are failing on Windows. @peastman would you have ideas on how to add the installation to Powershell? @nd-02110114 mentioned above that it wasn't clear how to get the Powershell install script working here

**Comment by nd-02110114** [comment]
> Would it make sense to represent indices as np.int32 instead of np.int64? It might improve performance.

I fixed it!

> I think our PyTorch Geometric tests are failing on Windows

The tests which failed on Windows CI are not related to PyTorch Geometric. The tests are related to `make_pytorch_dataset ` method of Dataset class. And then, the Powershell script is also not related to this error because it is not used on Windows CI. So, if you think the code is not problem, it is ready to merge.

**Comment by nd-02110114** [comment]
I simplified the install scripts and I finished fixing the Powershell script

**Comment by rbharath** [comment]
Does it look like there's a bug in `make_pytorch_dataset` that we're catching now because of the new tests?

CC @peastman

**Comment by nd-02110114** [comment]
The error results from starting to install PyTorch in CI. All `make_pytorch_dataset ` tests have been skipped before. Please check this line. I seem that  all `make_pytorch_dataset ` tests have not passed in Winddows since they were first added.

https://github.com/deepchem/deepchem/blob/b13f82de06f45c4ca515748a3a64635c2b91e0c3/deepchem/data/tests/test_datasets.py#L859 

So, I think the error is not directly related to this PR and I will try to fix the error by making an another PR.

**Comment by rbharath** [comment]
@nd-02110114 Yep, that sounds good! I think we have one merge conflict (just merged in @ncfrey's PR). Once the merge conflict is fixed, will go head and merge this in

**Comment by nd-02110114** [comment]
I resolved the conflicts!

**Comment by rbharath** [comment]
@nd-02110114 Ok sounds good!

@ncfrey @peastman As a quick heads up, this PR starts running tests for `make_pytorch_dataset` which were previously not running before on Windows. These are currently failing, but the failures don't look to be related to this PR, so I'm merging this in to block @nd-02110114. @nd-02110114 has kindly volunteered to fix the pytorch tests in a follow-up PR :)

**Comment by peastman** [comment]
Would you like me to try debugging the error with `make_pytorch_dataset()`?

**Comment by nd-02110114** [comment]
> Would you like me to try debugging the error with make_pytorch_dataset()?

Sorry for a late 🙇  I made PR #2082

**Pull request merged**